crates/data-plane-controller/src/controller.rs

```diff
@@ -157,16 +157,7 @@ impl Controller {
             self.validate_state(ops_checkout, state).await?;
         }
 
-        if !state.private_links.is_empty()
-            && !state.stack.config.model.private_links.is_empty()
-            && state.private_links != state.stack.config.model.private_links
-        {
-            anyhow::bail!(
-                "cannot set both config.private_links and private_links, prefer the latter."
-            );
-        }
-
-        state.stack.config.model.private_links = state.private_links.clone();
+        state.stack.config.model.private_links = row_state.stack.config.model.private_links.clone();
```

Comment:
Given the requirement that

> We want the _current_ state of the `private_links` column to be used for each individual `PulumiUp` attempt

This looks correct to me: we'll never get stuck trying a broken config over and over if we always take the value directly from the database, allowing updates to the broken config to get picked up on the next run. 

The caveat is that, unlike previously where we carried forward the previous `stats.private_links`, this will likely break any data plane that has existing private links configured via the `data_planes.config` column. So I think this will require a migration where we copy `config.private_links` into the `private_links` column.

... And now that I check the database, it looks like this has already been done.